### ISSUE TITLE
Use TensorDict to speed up data processing

### ISSUE THREAD

**Issue by taozhuo** [Unity-Technologies/ml-agents#5845]
**Is your feature request related to a problem? Please describe.**
Currently, simulation data from Unity has to go through many steps before the RL model consumes it:

RPC proto-> converted to `DecisionSteps, TerminalSteps(env)` -> converted to `Trajectory(AgentProcessor)` ->  put in`AgentManagerQueue[Trajector]` -> get `Trajectory(Trainer)` -> converted to `AgentBuffer` -> converted to `Torch Tensor` -> RL optimizer
 
There is a significant overhead in list operations during these conversions as well as serialization and deserialization costs from producer/consumer of AgentManagerQueue.

**Describe the solution you'd like**
`TensorDict` is a dictionary-like class that inherits properties from tensors, such as indexing, shape operations, and casting to device - we can run shuffle, sampling directly on a `TensorDict`. And it also supports shared memory access between multiple processes. By using the same copy of TensorDict we can save conversion costs and speed up processing, the dataflow will become:

RPC proto -> converted to `DecisionSteps, TerminalSteps(env)`  -> converted to `TensorDict` -> put in `Torch Queue` -> get `TensorDict`-> RL optimizer

A simple example of potential use of TensorDict:
```
batch_size = 2
tensordict1 = TensorDict(
    source={
        "obs": torch.zeros(batch_size, 1),
        "next_obs": torch.zeros(batch_size, 5, 6, dtype=torch.bool),
    },
    batch_size=[batch_size],
)

batch_size = 3
tensordict2 = TensorDict(
    source={
        "obs": torch.ones(batch_size, 1),
        "next_obs": torch.ones(batch_size, 5, 6, dtype=torch.bool),
    },
    batch_size=[batch_size],
)

tensordict = torch.cat([tensordict1, tensordict2])
tensordict.batch_size, tensordict["obs"]

Out[187]: 
(torch.Size([5]),
 tensor([[0.],
         [0.],
         [1.],
         [1.],
         [1.]]))

tensordict.share_memory_()
tensordict.is_shared()
Out[190]: True
```

**Additional context**
